nixos/k3s: streamline HA setup

yajo opened the pull request:
###### Description of changes


- Remove fake docs.
- Add new assertions to let configurations make more sense.
- Add clusterInit flag.
- Add some more docs about HA mode setup.

Fix https://github.com/NixOS/nixpkgs/issues/182085
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

@moduon MT-904

superherointj commented:
Please follow:

https://nixos.org/manual/nixos/unstable/#sec-writing-documentation

superherointj commented:
Notice CI errors on docs. Please follow:

https://nixos.org/manual/nixos/unstable/#sec-writing-documentation

superherointj commented on nixos/modules/services/cluster/k3s/default.nix:
```diff
@@ -33,15 +33,38 @@ in
+    clusterInit = mkOption {
+      type = types.bool;
+      default = cfg.role == "server";
+      example = false;
```
```suggestion
      defaultText = "server";
      example = false;
```

superherointj commented:
`Remove fake docs.` is bad wording.
Docs aren't fake. And you aren't removing docs. Please improve on it.

superherointj commented:
Would be nice if you also contributed a test to test the HA feature you are adding. Feel free to take ideas from [here](https://github.com/NixOS/nixpkgs/tree/master/nixos/tests/k3s).

pennae commented on nixos/modules/services/cluster/k3s/default.nix:
```diff
@@ -33,15 +33,38 @@ in
+        Servers and agents need to communicate each other.
+        Read [the networking docs][1] to know how to configure the firewall.
+
+        [1]: https://rancher.com/docs/k3s/latest/en/installation/installation-requirements/#networking
```
individual descriptions are not independent markdown instances, using references like this could lead to problems. (this is the first one to appear at all)

superherointj replied:
Maybe we could have an option to open firewall ports for K3s.

hall replied:
I'd like to see an `openFirewall` option here. If it helps, I've just recently setup an HA k3s cluster and these are the ports which had to be configured:
```nix
{
  networking.firewall = {
    allowedUDPPorts = [
      53 # dns
      8472 # cni (flannel vxlan)
    ];
    allowedTCPPorts = [
      10250 # metrics server
    ] ++ (lib.optionals (cfg.role == "server") [
      6443 # k8s api
      2379 # etcd client requests
      2380 # etcd peer communication
    ]);
  };
}
```

yajo replied:
I deliberately didn't add that option because it depends a lot on how you want to configure the firewall.

The linked docs explain that some ports you opened shouldn't be world-reachable. Instead, most ports should be only reachable among nodes. ⚠️ Beware if that's the setup you're deploying in production!

To properly set up the firewall connections, we need to know other nodes' IPs. That fits better in multi-node deployment tools such as nixops or similar.

So, for now, I'll leave the firewall setup to the sysadmin.

superherointj commented:
Please fix:
![image](https://user-images.githubusercontent.com/5861043/183435090-fdcf47df-2cf2-4184-8379-c93bd724cc24.png)

Mic92 commented on nixos/modules/services/cluster/k3s/default.nix:
```diff
@@ -33,15 +33,38 @@ in
+        The k3s server to connect to.
+
+        Servers and agents need to communicate each other.
+        Read [the networking docs][1] to know how to configure the firewall.
+
+        [1]: https://rancher.com/docs/k3s/latest/en/installation/installation-requirements/#networking
```
```suggestion
        Read (the networking docs)[https://rancher.com/docs/k3s/latest/en/installation/installation-requirements/#networking] to know how to configure the firewall.
```

superherointj commented:
@yajo Please update the PR. And report status.

yajo reviewed (commented):
Sorry for the late reply, I was on holidays.

All comments attended, except for the tests. I'll look into them now.

yajo commented:
Tests added.

superherointj commented on nixos/tests/k3s/multi-node.nix:
```diff
@@ -65,14 +65,38 @@ import ../make-test-python.nix ({ pkgs, ... }:
+        services.k3s = {
+          inherit tokenFile;
+          enable = true;
+          role = "server";
+          package = pkgs.k3s;
```
```suggestion
```
Already default values.

superherointj commented on nixos/tests/k3s/multi-node.nix:
```diff
@@ -65,14 +65,38 @@ import ../make-test-python.nix ({ pkgs, ... }:
+        networking.firewall.allowedUDPPorts = [ 8472 ];
+        networking.firewall.trustedInterfaces = [ "flannel.1" ];
+        networking.useDHCP = false;
+        networking.defaultGateway = "192.168.1.3";
```
```suggestion
        networking.defaultGateway = "192.168.1.3";
```
Why gateway is itself? In case this is not needed. It shouldn't be included then?

yajo replied:
I have no idea, I was just copying the other server definitions.

yajo replied:
I tried to remove this and the test failed, so it was needed after all.

superherointj commented on nixos/modules/services/cluster/k3s/default.nix:
```diff
@@ -33,15 +33,40 @@ in
 
+    clusterInit = mkOption {
+      type = types.bool;
+      default = cfg.role == "server";
```
```suggestion
      default = cfg.role == "server" && cfg.serverAddr == "";
```

superherointj replied:
It seems we have 3 roles: `master`, `replica`, `node`.
"master" won't have `serverAddr` but will have `clusterInit` true.
`replica` is a server but is not master, requires serverAddr and denies use of clusterInit.
Maybe we should change roles to reflect this automatically. And infer from there. Also warn when some field is missing.
`nodes` seek to use all IPs from master & replica in serverAddr.
Do you agree?

superherointj replied:
@euank Do you think `roles` option should account for HA needs?

euank replied:
I'm not totally sure.

The "roles" thing is there because the upstream docs and installation instructions talk a lot about "agent" vs "server".

Ideally, I'd like to keep our concepts as close to the upstream k3s docs so that those docs can help someone figure out how to use the nixos module too.

For consistency with [upstream docs](https://rancher.com/docs/k3s/latest/en/installation/ha-embedded/), I'd prefer to expose `clusterInit` over adding a new replica role to mask over it.

yajo replied:
I also think it's better to stick closer to upstream.

superherointj commented on nixos/modules/services/cluster/k3s/default.nix:
```diff
@@ -33,15 +33,40 @@ in
+    clusterInit = mkOption {
+      type = types.bool;
+      default = cfg.role == "server";
+      defaultText = literalExpression ''cfg.role == "server"'';
```
```suggestion
      defaultText = ''false";
```
Why is it necessary otherwise?

euank commented on nixos/modules/services/cluster/k3s/default.nix:
```diff
@@ -33,15 +33,40 @@ in
+
+        This option only makes sense in a server that is not connecting to another server.
+
+        If you are configuring an HA cluster, the 1st server must be configured
```
"an HA cluster" to "an HA cluster with an embedded DB" maybe? I know the doc comment's also already a bit wordy, but the other type of HA (external datastore) doesn't use the cluster-init flag at all.

euank commented on nixos/modules/services/cluster/k3s/default.nix:
```diff
@@ -33,15 +33,40 @@ in
+      description = lib.mdDoc ''
+        Initialize a cluster if k3s is started for the first time.
+
+        If the cluster is already initialized, this option has no effect.
```
From [the docs](https://rancher.com/docs/k3s/latest/en/installation/ha-embedded/#existing-clusters), it sounds like this does have an effect on existing clusters.

As it says:

> If you have an existing cluster using the default embedded SQLite database, you can convert it to etcd by simply restarting your K3s server with the --cluster-init flag

Since I think sqlite is / has-been the default, that means many existing clusters will be using it, and adding this default will migrate them forward to etcd.

The upstream default is still sqlite, and you have to opt-in to etcd, so I feel like that's the better default for us.

yajo replied:
I can expose the `--datastore-endpoint` option. Then assert that it is `etcd` if we're enabling cluster-init.

Finally I can change the defaults to be sqlite without cluster-init, to stick closer to upstream.

Does that seem a good solution?

yajo replied:
Reading https://rancher.com/docs/k3s/latest/en/installation/datastore/#external-datastore-configuration-parameters I can see that `--datastore-endpoint` shall be kept empty for any embedded DB usage.

It seems like k3s is smart enough to use embedded sqlite by default and to migrate to embedded etcd automatically when the server gets `--cluster-init` or when another server joins the 1st one.

So I'll just keep this option `false` by default and improve the docs to let users understand that this flag enables embedded etcd mode.

euank reviewed (commented):
Thanks for the PR!

I've got a question about defaulting on `--cluster-init` being the right default, but adding the ability to get a working HA w/ embedded etcd is great!

yajo committed:
k3s: streamline HA setup

- Replace misleading docs.
- Add new assertions to let configurations make more sense.
- Add clusterInit flag.
- Add some more docs about HA and non-HA modes setup.
- Improve multi-node tests for HA mode.

Fix https://github.com/NixOS/nixpkgs/issues/182085

yajo reviewed (commented):
Thanks everybody. All comments attended. I think the docs now state more clearly the different use cases.

Mic92 commented:
Thanks!

superherointj commented:
I was not done.
